daml-script/dump/src/main/scala/com/daml/script/dump/Encode.scala

```diff
@@ -0,0 +1,253 @@
+  }
+
+  private def qualifyId(id: Identifier): Doc =
+    Doc.text(id.moduleName) + Doc.text(".") + Doc.text(id.entityName)
+
+  private def encodeEv(
+      partyMap: Map[String, String],
+      cidMap: Map[String, String],
+      ev: TreeEvent.Kind,
+  ): Doc = ev match {
+    case TreeEvent.Kind.Created(created) =>
+      Doc.text("createCmd ") + encodeRecord(partyMap, cidMap, created.getCreateArguments)
+    case TreeEvent.Kind.Exercised(exercised @ _) =>
+      Doc.text("exerciseCmd ") + encodeCid(cidMap, exercised.contractId) + Doc.space + encodeValue(
+        partyMap,
+        cidMap,
+        exercised.getChoiceArgument.sum,
+      )
+    case TreeEvent.Kind.Empty => throw new IllegalArgumentException("Unknown tree event")
+  }
+
+  private def bindCid(cidMap: Map[String, String], c: CreatedContract): Doc = {
+    Doc.text("let ") + encodeCid(cidMap, c.cid) + Doc.text(" = createdCid @") +
+      qualifyId(c.tplId) + Doc.text(" [") + Doc.intercalate(
+        Doc.text(", "),
+        c.path.map(encodeSelector(_)),
+      ) + Doc.text("] tree")
+  }
+
+  private def encodeTree(
+      partyMap: Map[String, String],
+      cidMap: Map[String, String],
+      tree: TransactionTree,
+  ): Doc = {
+    val rootEvs = tree.rootEventIds.map(tree.eventsById(_).kind)
+    val submitters = rootEvs.flatMap(evParties(_)).toSet
+    val cids = treeCids(tree)
+    (Doc.text("tree <- submitTreeMulti ") + encodeParties(partyMap, submitters) + Doc.text(
+      " [] do"
+    ) /
+      Doc.stack(rootEvs.map(ev => encodeEv(partyMap, cidMap, ev)))).hang(2) /
+      Doc.stack(cids.map(bindCid(cidMap, _)))
+  }
+
+  private def encodeSelector(selector: Selector): Doc = Doc.str(selector.i)
+
+  private def encodeImport(moduleName: String) =
+    Doc.text("import qualified ") + Doc.text(moduleName)
+
+  private def partyMapping(parties: Set[String]): Map[String, String] = {
+    // - PartyIdStrings are strings that match the regexp ``[A-Za-z0-9:\-_ ]+``.
+    def safeParty(p: String) =
+      Seq(":", "-", "_", " ").foldLeft(p) { case (p, x) => p.replace(x, "") }.toLowerCase
+    // Map from original party id to DAML identifier
+    var partyMap: Map[String, String] = Map.empty
+    // Number of times we’ve gotten the same result from safeParty, we resolve collisions with a suffix.
+    var usedParties: Map[String, Int] = Map.empty
+    parties.foreach { p =>
+      val r = safeParty(p)
+      usedParties.get(r) match {
+        case None =>
+          partyMap += p -> s"${r}_0"
+          usedParties += r -> 0
+        case Some(value) =>
+          partyMap += p -> s"${r}_${value + 1}"
+          usedParties += r -> (value + 1)
+      }
+    }
+    partyMap
+  }
+
+  private def cidMapping(cids: Seq[Seq[CreatedContract]]): Map[String, String] = {
+    def lowerFirst(s: String) =
+      if (s.isEmpty) {
+        s
+      } else {
+        s.head.toLower.toString + s.tail
+      }
+    cids.zipWithIndex.toList.flatMap { case (cs, treeIndex) =>
```

Comment:
Should you use a view to create fewer intermediate copies?

Reply:
9deb2a234f521ee13c6eb9f08e108cb6853747cc